pricegraph/src/graph/shortest_paths.rs

```diff
@@ -124,15 +124,26 @@ where
         // each node: if a node can be relaxed then a negative cycle exists and is
         // created in the predecessor store; otherwise no such cycle exists.
 
-        for i in self.graph.node_identifiers() {
-            for edge in self.graph.edges(i) {
-                let j = edge.target();
-                let w = *edge.weight();
-
-                if *self.distance(i) + w < *self.distance(j) {
-                    self.update_predecessor(j, Some(i));
-                    return Some(j);
+        let mut visited: Vec<Option<usize>> = vec![None; self.predecessor_store.predecessors.len()];
+        for current_generation in 0..visited.len() {
+            if visited[current_generation].is_some() {
+                continue;
+            }
+            visited[current_generation] = Some(current_generation);
+            let mut current_index = current_generation;
+            while let Some(next) = self.predecessor_store.predecessors[current_index] {
+                let next_index = self.graph.to_index(next);
+                match visited[next_index] {
+                    Some(visited_generation) => {
+                        if visited_generation < current_generation {
```

Comment:
>  In particular, the check visited_generation < current_generation is perplexing. What is special about nodes that are after visited_generation?

@fedgiac visited generation can only be smaller or equal to current_generation. It it is smaller, it means we already visited this node in a previous path (and haven't found a cycle).
If it's equal it means we have been here in the current pass, thus we have found a cycle.

You are right that I should add some more comments into the logic.